apps/condo/domains/user/utils/googleRecaptcha3.js

```diff
@@ -1,45 +1,86 @@
+
+
 const GOOGLE_RECAPTCHA_CONFIG = conf.GOOGLE_RECAPTCHA_CONFIG ? JSON.parse(conf.GOOGLE_RECAPTCHA_CONFIG) : {}
-const CAPTCHA_SCORE_URL = GOOGLE_RECAPTCHA_CONFIG.CAPTCHA_SCORE_URL ? GOOGLE_RECAPTCHA_CONFIG.CAPTCHA_SCORE_URL : 'https://www.google.com/recaptcha/api/siteverify'
-const SERVER_KEY = GOOGLE_RECAPTCHA_CONFIG.SERVER_KEY
+const SITE_KEY = GOOGLE_RECAPTCHA_CONFIG.SITE_KEY
+const IOS_KEY = GOOGLE_RECAPTCHA_CONFIG.IOS_KEY
+const ANDROID_KEY = GOOGLE_RECAPTCHA_CONFIG.ANDROID_KEY
+const PROJECT_ID = GOOGLE_RECAPTCHA_CONFIG.PROJECT_ID
+const API_KEY = GOOGLE_RECAPTCHA_CONFIG.API_KEY
+const CAPTCHA_SCORE_URL = `https://recaptchaenterprise.googleapis.com/v1/projects/${PROJECT_ID}/assessments?key=${API_KEY}`
 const SAFE_CAPTCHA_SCORE = 0.5
-const THROW_ERRORS_ON_LOW_CAPTCHA_SCORE = false
+const THROW_ERRORS_ON_LOW_CAPTCHA_SCORE = conf.THROW_ERRORS_ON_LOW_CAPTCHA_SCORE === 'true'
+const DISABLE_CAPTCHA_CHECK = conf.DISABLE_CAPTCHA_CHECK === 'true'
 
 const logger = getLogger('googleRecaptcha3')
 
-if (isEmpty(SERVER_KEY)) {
+if (isEmpty(GOOGLE_RECAPTCHA_CONFIG) || !SITE_KEY || !IOS_KEY || !ANDROID_KEY || !PROJECT_ID || !API_KEY) {
+    console.error('Google reCaptcha not configured')
     logger.error('Google reCaptcha not configured')
 }
 
-const captchaCheck = async (response, action = '') => {
-    if (conf.NODE_ENV === 'test') {
+const getReCaptchaKey = (source) => {
+    switch (source) {
+        case REQUEST_SOURCES.ANDROID_APP:
+            return ANDROID_KEY
+        case REQUEST_SOURCES.IOS_APP:
+            return IOS_KEY
+        case REQUEST_SOURCES.SITE:
+            return SITE_KEY
+        default:
+            throw new Error('Unexpected request source to captcha check')
+    }
+}
+
+const captchaCheck = async (token, action = '', context = {}) => {
+    if (DISABLE_CAPTCHA_CHECK) {
         return { error: null }
     }
 
     try {
+        const source = getRequestSource(context)
+
         const serverAnswer = await fetch(CAPTCHA_SCORE_URL, {
             headers: {
-                'Content-Type': 'application/x-www-form-urlencoded',
+                'Content-Type': 'application/json; charset=utf-8',
             },
             method: 'POST',
-            body: `secret=${SERVER_KEY}&response=${response}`,
+            body: JSON.stringify({
+                event: {
+                    token: token,
+                    siteKey: getReCaptchaKey(source),
+                    expectedAction: action,
+                },
+            }),
         })
 
         const result = await serverAnswer.json()
         logger.info({ msg: 'Captcha result', result })
 
         if (serverAnswer.ok) {
-            const isError = THROW_ERRORS_ON_LOW_CAPTCHA_SCORE && result.score < SAFE_CAPTCHA_SCORE
-            if (result.action !== action) {
-                logger.error(`Captcha actions mismatch ${result.action} - ${action}`)
+            if (!THROW_ERRORS_ON_LOW_CAPTCHA_SCORE) return { error: null }
```

Comment:
why we need this logic?

Reply:
Looks like legacy to allow requests to pass without checking ratings, but still executing a request to verify the token.
I think we can leave this

Reply:
But I noticed that this check is very early, it should be just before the rating check